### PR TITLE
[showModalBottomSheet] fix: showModalBottomSheet does not move along keyboard

### PR DESCRIPTION
**Pull request by ZainUrRehmanKhan**
## Description

This issue was found in release 1.22, 1.23, 1.24 that the `showModelBottomSheet` was not moving along the keyboard, the bottom sheet contains a `TextField` which open up the keyboard on focus, so if the height of bottom sheet is less then the height of keyboard then the bottom sheet was hidden behind the keyboard.

**showModelBottomSheet with `TextField` unfocused:**

unfocus

**When we Focus on `TextField` - Before Changes:**
Bottom Sheet is hidden behind the keyboard.

before

**When we Focus on `TextField` - After Changes:**

after

## Related Issues

https://github.com/flutter/flutter/issues/71418

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by ZainUrRehmanKhan** [comment]
Kindly review my pull request.

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 1]
```diff
@@ -375,21 +375,24 @@ class _ModalBottomSheetState<T> extends State<_ModalBottomSheet<T>> {
...
+          },
+          builder: widget.route!.builder!,
+          backgroundColor: widget.backgroundColor,
+          elevation: widget.elevation,
+          shape: widget.shape,
+          clipBehavior: widget.clipBehavior,
+          enableDrag: widget.enableDrag,
+          onDragStart: handleDragStart,
+          onDragEnd: handleDragEnd,
+        )
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/bottom_sheet.dart:395]:
```suggestion
        ),
```

**Review by dkwingsmt — CHANGES_REQUESTED** [review]
The change looks solid. Please add a unit test to make sure someone doesn't accidentally revert it (probably a [golden test](https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package%3Aflutter)).

**Comment by ZainUrRehmanKhan** [comment]
I'm writing a test for this PR.

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 2]
```diff
@@ -726,6 +726,41 @@ void main() {
 
     expect(retrievedRouteSettings, routeSettings);
   });
+
+  testWidgets('showModalBottomSheet should move along keyboard',
```
Comment (dkwingsmt) [inline packages/flutter/test/material/bottom_sheet_test.dart:730]:
```suggestion
  testWidgets('showModalBottomSheet should move along on-screen keyboard',
```

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 3]
```diff
@@ -726,6 +726,41 @@ void main() {
...
+          (WidgetTester tester) async {
+    late BuildContext savedContext;
+
+    //Wrapping Builder with MediaQuery for KeyboardHeight
```
Comment (dkwingsmt) [inline packages/flutter/test/material/bottom_sheet_test.dart:734]:
```suggestion
    // Show a keyboard (simulate by space at the bottom of the screen).
```

**Review by dkwingsmt — CHANGES_REQUESTED** [review]
Tiny issues. Everything else looks great!

**Comment by ZainUrRehmanKhan** [comment]
Suggested changes done.

**Comment by ZainUrRehmanKhan** [comment]
@dkwingsmt Updated my branch to the latest master. That should resolve the failing customer tests.

**Comment by ZainUrRehmanKhan** [comment]
?

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Comment by davidmartos96** [comment]
@dkwingsmt Could we have a way to make this change opt out? I currently have some modal bottom sheets where I place the textfields (usually for filtering results) at the top of a quite large bottomSheet, so the keyboard will not hide it. With this PR everything will be pushed above the keyboard which may be considered somewhat breaking. What are your thoughts?

**Comment by dkwingsmt** [comment]
@davidmartos96 You are right. Although to make a modal bottom sheet hidden by the on-screen keyboard does not really sound correct to me. When a displayed modal bottom sheet is supposed to prevent the user from interacting with the rest of the app. Will normal `showBottomSheet` suit your needs?

**Comment by davidmartos96** [comment]
@dkwingsmt I may have not expressed myself correctly. Sorry about that.
The usecase I'm referring to is a TextField inside the modal bottom sheet, but where the bottom sheet will be about 70% tall of the screen height. I have some of those modal sheets in my app, but I display the text field at the top knowing that up to now the keyboard would hide the lower portion of the bottom sheet (and I'm fine with that in my usecase). 
With the PR I worry that all the bottom sheet will be pushed too much overflowing the screen height. I agree with the changes made in the PR but I would suggest including an option similar to `resizeToAvoidBottomInset` in the Scaffold so that this behavior can be opted out (https://api.flutter.dev/flutter/material/Scaffold/resizeToAvoidBottomInset.html)
If the Flutter team agrees to such new option I could open a PR with it.

**Comment by dkwingsmt** [comment]
Can you achieve the same result by manually applying a negative bottom padding? Actually, you can get the height of your sheet and the screen height to calculate the exact amount of underflow for you to avoid overflowing.

**Comment by davidmartos96** [comment]
@dkwingsmt I'm not sure I follow you with the negative padding. Does Flutter support negative padding?

**Comment by Peng-Qian** [comment]
Personally, I think this is a bad idea to force ModalBottomSheet to resize itself. I am experiencing the same issue as @davidmartos96 says.

At least we need a parameter to control if the ModalBottomSheet should resize or not (). And why padding should wrap BottomSheet to force resize to avoid keyboard? (we can just add padding inside ModalBottomSheet to achieve the same effect.)

**Comment by dkwingsmt** [comment]
@HansMuller Are you more familiar with how Material Design defines ModelBottomSheet? Can you take a look at the requests above?

**Comment by Peng-Qian** [comment]
Thanks for your quick response!

Please imagine ModalBottomSheet takes 5/6 of full-screen height and it used to pick a country(e.g. New Zealand), so it will be a scrollable list and contains many items, and on the top of this list, there is a search bar that contains a TextField.

If the ModalBottonSheet forcibly resizes as the keyboard shows, It will push the search bar to the very top of the screen, it not only breaks the UI looking but also break the feature. Because the phone notification bar will overlay on it.

Moreover, padding when the keyboard shows is a brutal way to avoid keyboard overlay. Please check out this https://pub.dev/packages/keyboard_avoider package. I don't think flutter should limit us on how to design the Application.

**Comment by dkwingsmt** [comment]
We're reverting this PR in https://github.com/flutter/flutter/pull/77286. Various sources have reported that the previous behavior is the desired one. /cc @ZainUrRehmanKhan

**Comment by totroi** [comment]
> We're reverting this PR in #77286. Various sources have reported that the previous behavior is the desired one.

Any tips on what to do to have the behavior of the original commit (moving up with the keyboard)?

Maybe an specific modal could be created, like _showModalBottomSheetFlexible_ ? I'd prefer to have it behave that way

**Comment by totroi** [comment]
> > We're reverting this PR in #77286. Various sources have reported that the previous behavior is the desired one.
> 
> Any tips on what to do to have the behavior of the original commit (moving up with the keyboard)?
> 
> Maybe an specific modal could be created, like _showModalBottomSheetFlexible_ ? I'd prefer to have it behave that way

Actually it you be much better if one could simply define the "height" of it. Being able to elongate it up just a little would solve my problem.

**Comment by totroi** [comment]
> @totroi You can get the behavior by wrapping the bottom sheet builder with:
> 
> ```dart
> Padding(
>   padding: MediaQuery.of(context).viewInsets
> )
> ```
> 
> That's essentially what the reverted PR was doing

Thanks! Problem is I'm seeking to move a Card inside the modal up, and move the modal up just the amount required for the Card to move. Like about 30 pixels...

I managed to shrink all the paddings inside the Card widget and it looks okay, but the best result would be those 30 pixels up whenever the keyboard is opened.

**Pull request merged**